### PR TITLE
RavenDB-21734 - Support Direct Upload for Azure

### PR DESCRIPTION
**Pull request by grisha-kotler**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-21734/Support-Direct-Upload-for-Azure

### Additional description

Support Direct Upload for Azure.

### Type of change

- Optimization
- New feature

### How risky is the change?

- Low 

### Backward compatibility

- Non breaking change

### Documentation update

- This change requires a documentation update. Please mark the issue on YouTrack using `Documentation Required` tag.

### Testing by Contributor

- Tests have been added that prove the fix is effective or that the feature works
- It has been verified by manual testing

### Testing by RavenDB QA team

- No special testing by RavenDB QA team is needed

### UI work

- No UI work is needed

### CODE REVIEW

File: src/Raven.Server/Documents/PeriodicBackup/Azure/RavenAzureClient.cs [hunk 1]
```diff
@@ -46,7 +47,7 @@ public class RavenAzureClient : IProgress<long>, IRavenAzureClient
 
         public Size MaxSingleBlockSize { get; set; } = new Size(256, SizeUnit.Megabytes);
 
-        private RavenAzureClient(AzureSettings azureSettings, BackupConfiguration configuration, Progress progress = null, CancellationToken cancellationToken = default)
+        public RavenAzureClient(AzureSettings azureSettings, BackupConfiguration configuration, Progress progress = null, CancellationToken cancellationToken = default)
```
Comment (ppekrol) [inline src/Raven.Server/Documents/PeriodicBackup/Azure/RavenAzureClient.cs]: private

File: src/Raven.Server/Documents/PeriodicBackup/DirectUpload/DirectUploadBackupTask.cs [hunk 2]
```diff
@@ -27,26 +28,41 @@ protected override Stream GetStreamForBackupDestination(string filePath, string
...
+                return new AwsS3DirectUploadStream(GetDirectUploadParameters(
+                    progress => new RavenAwsS3Client(s3Settings, Database.Configuration.Backup, progress, TaskCancelToken.Token),
+                    s3Settings.RemoteFolderName, folderName, fileName));
+
+            case BackupConfiguration.BackupDestination.Azure:
+                var azureSettings = GetBackupConfigurationFromScript(Configuration.AzureSettings, x => JsonDeserializationServer.AzureSettings(x),
+                    settings => PutServerWideBackupConfigurationCommand.UpdateSettingsForAzure(settings, Database.Name));
+
+                return new AzureDirectUploadStream(GetDirectUploadParameters(
+                    progress => new RavenAzureClient(azureSettings, Database.Configuration.Backup, progress, TaskCancelToken.Token),
```
Comment (ppekrol) [inline src/Raven.Server/Documents/PeriodicBackup/DirectUpload/DirectUploadBackupTask.cs]: use factory Create?

File: src/Raven.Server/Documents/PeriodicBackup/DirectUpload/AzureMultiPartUploader.cs [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    {
+        AsyncHelpers.RunSync(() => UploadPartAsync(stream, size));
+    }
+
+    public async Task UploadPartAsync(Stream stream, long size)
+    {
+        var blockID = Convert.ToBase64String(BitConverter.GetBytes(_partNumber++));
+        _base64BlockIds.Add(blockID);
+
+        stream.SetLength(size); // there is no upload size parameter that we can use in the Azure API (unlike AWS)
```
Comment (ppekrol) [inline src/Raven.Server/Documents/PeriodicBackup/DirectUpload/AzureMultiPartUploader.cs]: isn't that DirectUpload stream here?

File: src/Raven.Server/Documents/PeriodicBackup/Azure/RavenAzureClient.cs (hunk 1 above)
Comment (grisha-kotler) [reply src/Raven.Server/Documents/PeriodicBackup/Azure/RavenAzureClient.cs]: done

File: src/Raven.Server/Documents/PeriodicBackup/DirectUpload/DirectUploadBackupTask.cs (hunk 2 above)
Comment (grisha-kotler) [reply src/Raven.Server/Documents/PeriodicBackup/DirectUpload/DirectUploadBackupTask.cs]: done

File: src/Raven.Server/Documents/PeriodicBackup/DirectUpload/AzureMultiPartUploader.cs (hunk 3 above)
Comment (grisha-kotler) [reply src/Raven.Server/Documents/PeriodicBackup/DirectUpload/AzureMultiPartUploader.cs]: This is a `MemoryStream`.

**Review by ppekrol — APPROVED** [review]

**Pull request merged**